### ISSUE TITLE
Llama4 Memory Leak + Convergence Inconsitencies with TP

### ISSUE THREAD

**Issue by lckr** [pytorch/torchtitan#2058]
### Bug description

Hi all! I was trying to reproduce the convergence experiments for llama4 as detailed [here](https://github.com/pytorch/torchtitan/blob/main/docs/converging.md#guidelines).

Looking at the resulting training statistics I've made two observations about potential issues with TP.

1. with `TP=2` there seems to be a memory leak which can lead to training crashes
    - This does not appear with `TP=4`

<img width="420" height="356" alt="Image" src="https://github.com/user-attachments/assets/19138220-e8fb-4cbb-9efd-7bdcb5deb140" />

2. While the runs `dp=2` and `dp=2,tp=2` are consistent (as expected), `dp=2,tp=4` is not (unexpected).

<img width="426" height="369" alt="Image" src="https://github.com/user-attachments/assets/486798e4-672c-409a-93cb-c051f98da89e" />

| Parallelism     | Matches Baseline | Memory Leak |
|-----------------|------------------|-------------|
| DP=2 (baseline) | +                | -           |
| DP=2,TP=2       | +                | +           |
| DP=2,TP=4       | -                | -           |
|                 |                  |             |
| DP=4 (baseline) | +                | -           |
| DP=4,TP=2       | +                | +           |

**Note:** I also tried the same experiment using the `llama3` debug model. The llama3 results do not show the above inconsistencies / memory leak but match the consistent train loss curves as in the [docs](https://github.com/pytorch/torchtitan/blob/main/docs/converging.md#test-results)

## Experiment Details:
- Llama4 debug model
- 1k steps on `c4` at `global_batch_size = 64` with `seq_len = 2048`
- `--compile.enable`
- No activation checkpointing
- Using seed checkpoint created with `NGPU=1 CONFIG_FILE="./torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.create_seed_checkpoint --parallelism.data_parallel_replicate_degree 1 --parallelism.data_parallel_shard_degree 1 --parallelism.tensor_parallel_degree 1 --parallelism.pipeline_parallel_degree 1 --parallelism.context_parallel_degree 1 --parallelism.expert_parallel_degree 1`
- additional debug flags:
```
--debug.moe_force_load_balance
--debug.deterministic
--debug.seed 42
```
- fixed global and local batch size
- Runs executed on a 8xH100 node

### Versions
- `torch`: `2.10.0.dev20251114+cu128`
- `torchao`: `0.15.0.dev20251119+cu12`
- `torchtitan`: current main at [bfdc97](https://github.com/pytorch/torchtitan/commit/bfdc974f94e033999362e2bf0ee0842457d43470)

<details>
<summary> Full Training Commands</summary>

```
NGPU=4 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 4 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance

NGPU=2 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 2 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance
  
NGPU=2 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 1 --parallelism.tensor_parallel_degree 2 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance

NGPU=4 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 1 --parallelism.tensor_parallel_degree 4 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance

NGPU=4 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance

NGPU=8 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 4 --parallelism.tensor_parallel_degree 2 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance

NGPU=8 CONFIG_FILE="torchtitan/models/llama4/train_configs/debug_model.toml" ./run_train.sh --checkpoint.enable --checkpoint.load_only --metrics.enable_wandb --metrics.log_freq 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 4 --debug.deterministic --debug.seed 42 --training.dataset c4 --lr_scheduler.warmup_steps 200 --training.steps 1000 --compile.enable --activation_checkpoint.mode none --training.global_batch_size 64  --debug.moe_force_load_balance
```

### Versions

- `torch`: `2.10.0.dev20251114+cu128`
- `torchao`: `0.15.0.dev20251119+cu12`
- `torchtitan`: current main at [bfdc97](https://github.com/pytorch/torchtitan/commit/bfdc974f94e033999362e2bf0ee0842457d43470)

**Comment by lckr** [comment]
I also repeated the runs with 2 layers of the llama4 arch
(again with all determinism setting + seed checkpoint)
```
"17bx16e_debug": TransformerModelArgs(
        dim=5120,
        n_layers=2,
        n_heads=40,
        n_kv_heads=8,
        ffn_dim_multiplier=1.2,
        multiple_of=2048,
        rope_theta=500000,
        rope_scaling_args=RoPEScalingArgs(),
        max_seq_len=10485760,
        moe_args=MoEArgs(num_experts=16),
        interleave_moe_layer_step=1,
    )
```

Results:
- Memory leak cannot be observed in any settings (my guess here is that this could be shape related)

- Inconsistency between `dp=2`, `dp=2,tp=2` and `dp=2,tp=4` persist

**Comment by tianyu-l** [comment]
Thanks for reporting the issue. Does it happen without `compile.enabled`?

**Comment by lckr** [comment]
Hi, 

Reran the experiments on llama4 debug model with `--compile.no_enable`.

The memory leak could not reproduced that way, but the inconsistency with TP still persists:

<img width="1085" height="858" alt="Image" src="https://github.com/user-attachments/assets/063f7afa-862d-4367-8c40-1c0701a8fa56" />

**Comment by tianyu-l** [comment]
> The memory leak could not reproduced that way, but the inconsistency with TP still persists:

Are you using the same seed checkpoint? It makes sure the init are the same across runs.
https://github.com/pytorch/torchtitan/blob/main/docs/debugging.md#reproducibility-between-runs

Could you try `mixed_precision_param = "float32"` + "Adam" optimizer? In the past I've found AdamW to give different numerics results when TP is used, but I haven't followed up.

cc @xmfan on the memory leakage issue under compile.

**Comment by lckr** [comment]
> Are you using the same seed checkpoint? It makes sure the init are the same across runs.

Yeah I used the same seed checkpoints within an experiment. But the seed checkpoints between the experiments are not the same.

> Could you try `mixed_precision_param = "float32"` + "Adam" optimizer

Will do and report back

**Comment by tianyu-l** [comment]
pls reopen if the issue persists
